feat: add support for creating a Matrix Factorization model

rey-esp opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

tswast committed:
docs: update title of pypi notebook example to reflect use of the PyPI public dataset

In response to feedback on internal change 662899733.

rey-esp committed:
feat: add support for creating a  Matrix Factorization model

rey-esp committed:
feat: add support for creating a  Matrix Factorization model

rey-esp committed:
feat: add support for creating a  Matrix Factorization model

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +198,177 @@ def score(
+        self,
+        n_components: Optional[Union[int, float]] = None,
+        *,
+        num_factors: int,
```
Allow `None` and keep the default as `None` so that the auto-detection can be supported (see: https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-matrix-factorization#num_factors).

```suggestion
        num_factors: Optional[int] = None,
```

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +198,177 @@ def score(
+        num_factors: int,
+        user_col: str,
+        item_col: str,
+        l2_reg: float,
```
Per: https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-matrix-factorization#l2_reg

* The default value is 1.0.
* This could also be a `HPARAM_RANGE(1.5, 5.0)` or `HPARAM_CANDIDATES([0, 1.0, 3.0, 5.0])`. Let's add a TODO to support hyperparameter tuning. (Maybe in a subsequent PR).

```suggestion
        # TODO: Add support for hyperparameter tuning.
        l2_reg: float = 1.0,
```

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
rating_col

rey-esp committed:
(nearly) complete class

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
removem print()

rey-esp committed:
removem print()

rey-esp committed:
adding recommend

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,93 @@
+            The item column name.
+        l2_reg (float, default 1.0):
+            If you aren't running hyperparameter tuning, then you can specify a FLOAT64 value. The default value is 1.0.
+            If you are running hyperparameter tuning, then you can use one of the following options:
+                The HPARAM_RANGE keyword and two FLOAT64 values that define the range to use for the hyperparameter. For example, L2_REG = HPARAM_RANGE(1.5, 5.0).
+                The HPARAM_CANDIDATES keyword and an array of FLOAT64 values that provide discrete values to use for the hyperparameter. For example, L2_REG = HPARAM_CANDIDATES([0, 1.0, 3.0, 5.0]).
```
The command `nox -r -s docs` is failing. I believe it's because of these lines.

```
sphinx.errors.SphinxWarning: /home/runner/work/python-bigquery-dataframes/python-bigquery-dataframes/bigframes/ml/decomposition.py:docstring of bigframes.ml.decomposition.MatrixFactorization:21:Unexpected indentation.
```

Since we aren't supporting hyperparameter tuning yet (internal issue 384510566), please remove references to these features.

```suggestion
```

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,93 @@
+        item_col (str):
+            The item column name.
+        l2_reg (float, default 1.0):
+            If you aren't running hyperparameter tuning, then you can specify a FLOAT64 value. The default value is 1.0.
```
Same here, remove references to hyperparameter tuning.

```suggestion
            A floating point value for L2 regularization. The default value is 1.0.
```

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,93 @@
+    Args:
+        num_factors (int or auto, default auto):
+            Specifies the number of latent factors to use.
+            If you aren't running hyperparameter tuning, then you can specify an INT64 value between 2 and 200. The default value is log2(n), where n is the number of training examples.
```
Remove references to hyperparameter tuning for now.

```suggestion
            Specify an integer value between 2 and 200.
            The default value is log2(n), where n is the number of training examples.
```

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
remove hyper parameter runing references

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

sycai reviewed (dismissed):
Does it make sense to add some system/unit tests for this change?

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
swap predict in _mf for recommend

rey-esp committed:
recommend -> predict

rey-esp committed:
update predict doc string

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +198,113 @@ def score(
+        *,
+        feedback_type: Literal["explicit", "implicit"] = "explicit",
+        num_factors: int,
+        user_col: str,
+        item_col: str,
+        rating_col: str = "rating",
```
@GarrettWu @shuoweil I see in https://github.com/googleapis/python-bigquery-dataframes/pull/1282 you ended up passing in "id_col" as a separate argument to `fit()` instead of the class constructor. Is this a pattern you would recommend here?

Note: MatrixFactorization differs somewhat from that application in that normally in scikit-learn one would have a "sparse matrix" data type (e.g. https://docs.scipy.org/doc/scipy/reference/sparse.html) where rows/cols/values would all be bundled up in one object, similar to how we are using the bigframes DataFrame for this purpose.

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
preparing test files

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
add test data

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
new error: to_gbq column names need to be changed?

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
Delete demo.ipynb

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
passing system test

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
preparing to add unit tests

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
2 out of 3 (so far) passing unit tests

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
attempted mocking

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
fix tests

rey-esp committed:
new test file for model creation unit tests

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
add unit tests for num_factors, user_col, and item_col

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

tswast commented on tests/unit/ml/test_matrix_factorization.py:
```diff
@@ -0,0 +1,100 @@
+#     num_factors = 0.5
+#     with pytest.raises(TypeError):
+#         decomposition.MatrixFactorization(
+#             num_factors=num_factors,
```
I suspect you weren't able to commit because the type checker was kicking in. When we intentionally pass in the wrong type, we can add a `  # type: ignore` comment.


```suggestion
#             # Intentionally pass in the wrong type. This will fail if the user is using
#             # a type checker, but we can't assume that everyone is doing so, especially
#             # not in notebook environments.
#             num_factors=num_factors,  # type: ignore
```

tswast commented on tests/unit/ml/test_matrix_factorization.py:
```diff
@@ -0,0 +1,100 @@
+
+# def test_decomposition_mf_invalid_num_factors_raises():
+#     num_factors = 0.5
+#     with pytest.raises(TypeError):
```
Would be good to make this a little more robust and make sure the error corresponds to the `num_factors` argument.

```suggestion
#     with pytest.raises(TypeError, match="num_factors"):
```

rey-esp committed:
Update tests/unit/ml/test_matrix_factorization.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update tests/unit/ml/test_matrix_factorization.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
uncomment one test

rey-esp committed:
uncomment test

rey-esp committed:
uncomment test

rey-esp committed:
uncomment test

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
nearly all tests

rey-esp committed:
tests complete and passing

rey-esp committed:
seeing if test causes kokoro failure

rey-esp committed:
uncomment test-kokoro still failing

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'tswast-patch-1' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
remove comment

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
fix test

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
test kokoro

rey-esp committed:
test_decomposition.py failing and now feedback_type attr does not exist

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
passing tests

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,163 @@ def score(
+        l2_reg: float = 1.0,
+    ):
+
+        if feedback_type not in ("explicit", "implicit", "EXPLICIT", "IMPLICIT"):
+            raise ValueError(
+                f"Expected feedback_type to be `explicit` or `implicit`, but got {feedback_type}"
+            )
```
I would suggest making this case-insensitive, instead.

```suggestion
        feedback_type = feedback_type.lower()
        if feedback_type not in ("explicit", "implicit"):
            raise ValueError(
                f"Expected feedback_type to be `explicit` or `implicit`, but got {feedback_type}"
            )
```

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Update tests/system/large/ml/test_decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
doc attempt - _mf.py example

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
feedback_type case ignore

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,91 @@
+        >>> X = bpd.DataFrame([[1, 1], [2, 1], [3, 1.2], [4, 1], [5, 0.8], [6, 1]])
+        >>> model = MatrixFactorization(feedback_type='explicit', num_factors=6, user_col='user_id', item_col='item_col', l2_reg=2.06)
+        >>> W = model.fit(X)
+        >>> result = model.global_explain()
```
Remove this reference?

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Update _mf.py - remove global_explain()

rey-esp committed:
fit

rey-esp committed:
pull?

rey-esp committed:
W

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
fix docs (maybe)

rey-esp committed:
Update test_matrix_factorization.py with updated error messages

rey-esp committed:
ilnt

rey-esp committed:
Update test_matrix_factorization.py - add 'f'

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,158 @@ def score(
+
+        self.rating_col = rating_col
+
+        if type(l2_reg) is not float:
```
This is probably a bit too precise. We should allow integer here.

Try https://docs.python.org/3/library/numbers.html#numbers.Real instead.

rey-esp committed:
improve errors and update tests

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

sycai reviewed (commented):
The extensive test coverage LGTM!

I will leave the approval to @tswast

tswast commented:
`doctest` failures appear related:

```
FAILED third_party/bigframes_vendored/sklearn/decomposition/_mf.py::_mf.MatrixFactorization
FAILED bigframes/ml/decomposition.py::bigframes.ml.decomposition.MatrixFactorization
2 failed, 640 passed, 140 warnings in 1952.43s (0:32:32)
```

```
UNEXPECTED EXCEPTION: BadRequest('GET [https://bigquery.googleapis.com/bigquery/v2/projects/bigframes-testing/queries/01104370-288e-4440-8653-680699149e53?maxResults=0&location=US&prettyPrint=false](https://www.google.com/url?q=https://bigquery.googleapis.com/bigquery/v2/projects/bigframes-testing/queries/01104370-288e-4440-8653-680699149e53?maxResults%3D0%26location%3DUS%26prettyPrint%3Dfalse&sa=D): Input data in query statement for matrix factorization is missing columns. Expected 3 received 2.')
```

on both.

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -27,7 +28,15 @@
+    "num_factors": "numFactors",
+    "user_col": "userColumn",
+    "item_col": "itemColumn",
+    # TODO: Add rating_col
```
Are we blocked on something for this TODO? If so, please add context. If not, let's add support for this.

Edit: Look at https://cloud.google.com/bigquery/docs/reference/rest/v2/models#trainingoptions for the corresponding parameters.

Edit2: I found userColumn and itemColumn but nothing corresponding to rating column. I think this is similar to that lower bounds issue (bug 391399223).

Edit3: I found in the code that `rating_col` gets written to `inputLabelColumns`, which will be a list of columns (just one if set in this case). I think we might need to do something with `@property` to solve this difference. Something like the following:

```suggestion
    "_input_label_columns": "inputLabelColumns",
```

and then below a `@property` for `rating_col` that looks at `_input_label_columns` and returns the first one if set, otherwise `None`.

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+
+        self.l2_reg = l2_reg
+        self._bqml_model: Optional[core.BqmlModel] = None
+        self._bqml_model_factory = globals.bqml_model_factory()
```
```suggestion
        self._bqml_model_factory = globals.bqml_model_factory()

    @property
    def rating_col(self) -> str:
        """str: The rating column name. Defaults to 'rating'.""""
        return self._input_label_columns[0]
```

tswast commented on tests/system/large/ml/test_decomposition.py:
```diff
@@ -163,3 +163,54 @@ def test_decomposition_configure_fit_load_none_component(
+    assert reloaded_model.num_factors == 6
+    assert reloaded_model.user_col == "user_id"
+    assert reloaded_model.item_col == "item_col"
+    assert reloaded_model.rating_col == "rating"
```
Should be `ratings` once the `@property` is implemented.

```suggestion
    assert reloaded_model.rating_col == "ratings"
```

tswast commented on tests/unit/ml/test_golden_sql.py:
```diff
@@ -209,3 +210,55 @@ def test_logistic_regression_score(mock_session, bqml_model, mock_X, mock_y):
+
+    mock_session.read_gbq.assert_called_once_with(
+        "SELECT * FROM ML.RECOMMEND(MODEL `model_project`.`model_dataset`.`model_id`,\n  (input_X_sql))",
+        index_col=["index_column_id"],
```
Why is `index_col` here? I only see `trial_id` in https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-recommend?hl=en

tswast commented on tests/unit/ml/test_matrix_factorization.py:
```diff
@@ -0,0 +1,213 @@
+    user_col = "user_col"
+    with pytest.raises(
+        ValueError,
+        match="Expected user_col column to be `user_id`.",
```
I'm a bit confused by this one. Shouldn't we support any column name?

rey-esp replied:
I kept hitting errors when I tried to create a model that didn't have a specific name. I'll go through it again and see if I still run into the problem - I may be confusing it in some way

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+
+        if not isinstance(item_col, str):
+            raise TypeError(f"Expected item_col to be STR, but got {type(item_col)}.")
+
+        if item_col != "item_col":
+            raise ValueError("Expected item_col column to be `item_col`.")
```
Please remove this validation. We should support other column names besides "item_col".

```suggestion
```

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+
+        if not isinstance(user_col, str):
+            raise TypeError(f"Expected user_col to be STR, but got {type(user_col)}.")
+
+        if user_col != "user_id":
+            raise ValueError("Expected user_col column to be `user_id`.")
```
Please remove this validation. We should support other column names besides "user_id".

```suggestion
```

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+
+        if not isinstance(num_factors, int):
+            raise TypeError(
+                f"Expected num_factors to be INT, but got {type(num_factors)}."
```
Let's use the Python naming for the types.

```suggestion
                f"Expected num_factors to be an int, but got {type(num_factors)}."
```

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+        self.num_factors = num_factors
+
+        if not isinstance(user_col, str):
+            raise TypeError(f"Expected user_col to be STR, but got {type(user_col)}.")
```
```suggestion
            raise TypeError(f"Expected user_col to be a str, but got {type(user_col)}.")
```

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+
+        if not isinstance(rating_col, str):
+            raise TypeError(
+                f"Expected rating_col to be STR, but got {type(rating_col)}."
```
```suggestion
                f"Expected rating_col to be a str, but got {type(rating_col)}."
```

tswast commented on bigframes/ml/decomposition.py:
```diff
@@ -197,3 +206,160 @@ def score(
+
+        if not isinstance(l2_reg, (float, int)):
+            raise TypeError(
+                f"Expected l2_reg to be FLOAT or INT, but got {type(l2_reg)}."
```
```suggestion
                f"Expected l2_reg to be a float or int, but got {type(l2_reg)}."
```

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,90 @@
+
+        >>> import bigframes.pandas as bpd
+        >>> from bigframes.ml.decomposition import MatrixFactorization
+        >>> X = bpd.DataFrame([[1, 1], [2, 1], [3, 1.2], [4, 1], [5, 0.8], [6, 1]])
```
We need the cell based decomposition e.g. (row, column, value) version of a matrix to be usable in BigQuery ML and bigframes.

Let's try that. This DataFrame would be decomposed to the following:

```suggestion
        >>> X = bpd.DataFrame({
        ... "row": [0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6],
        ... "column": [0, 1] * 6,
        ... "value": [1, 1, 2, 1, 3, 1.2, 4, 1, 5, 0.8, 6, 1],
        ... })
```

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,90 @@
+        >>> import bigframes.pandas as bpd
+        >>> from bigframes.ml.decomposition import MatrixFactorization
+        >>> X = bpd.DataFrame([[1, 1], [2, 1], [3, 1.2], [4, 1], [5, 0.8], [6, 1]])
+        >>> model = MatrixFactorization(feedback_type='explicit', num_factors=6, user_col='user_id', item_col='item_col', l2_reg=2.06)
```
```suggestion
        >>> model = MatrixFactorization(feedback_type='explicit', num_factors=6, user_col='row', item_col='column', rating_col='value', l2_reg=2.06)
```

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,90 @@
+        >>> from bigframes.ml.decomposition import MatrixFactorization
+        >>> X = bpd.DataFrame([[1, 1], [2, 1], [3, 1.2], [4, 1], [5, 0.8], [6, 1]])
+        >>> model = MatrixFactorization(feedback_type='explicit', num_factors=6, user_col='user_id', item_col='item_col', l2_reg=2.06)
+        >>> W = model.fit(X.rename(columns={0:'user_id', 2: 'rating_col', 1: 'item_col'}))
```
```suggestion
        >>> W = model.fit(X)
```

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Update tests/system/large/ml/test_decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py - num_factors error messsage

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py - user_col error message

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py - rating_col error message

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py - l2_reg error msg

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
fix tests to match updated error messages

rey-esp committed:
Merge branch 'b338873783-matrix-factorization' of github.com:googleapis/python-bigquery-dataframes into b338873783-matrix-factorization

rey-esp committed:
Update third_party/bigframes_vendored/sklearn/decomposition/_mf.py - docs df

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update third_party/bigframes_vendored/sklearn/decomposition/_mf.py - docs model

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update third_party/bigframes_vendored/sklearn/decomposition/_mf.py - docs fit

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

tswast commented on third_party/bigframes_vendored/sklearn/decomposition/_mf.py:
```diff
@@ -0,0 +1,94 @@
+        >>> from bigframes.ml.decomposition import MatrixFactorization
+        >>> X = bpd.DataFrame({
+        ... "row": [0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6],
+        ... "column": [0, 1] * 6,
```
Re: doctest failure for `UNEXPECTED EXCEPTION: ValueError('All arrays must be of the same length')`, I counted wrong in my previous suggestion. This should fix it!

```suggestion
        ... "column": [0, 1] * 7,
```

tswast committed:
Update third_party/bigframes_vendored/sklearn/decomposition/_mf.py

tswast commented:
Re: e2e error

```
>       assert reloaded_model.rating_col == "ratings"
E       AssertionError: assert 'rating' == 'ratings'
```

This is the correct assertion. Adding a `rating_col` `@property` to the `MatrixFactorization` class should fix it.

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
remove errors and tests

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Update bigframes/ml/decomposition.py

Co-authored-by: Tim Sweña (Swast) <swast@google.com>

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
passing system test

rey-esp committed:
E           AssertionError: expected call not found.
E           Expected: read_gbq('SELECT * FROM ML.RECOMMEND(MODEL ..,\n  (input_X_sql))', trial_id=['index_column_id'])
E           Actual: read_gbq('SELECT * FROM ML.RECOMMEND(MODEL ..,\n  (input_X_sql))', index_col=['index_column_id'])

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
same # of elements in each

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
attempt

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

rey-esp committed:
doc fix

rey-esp committed:
doc fix

rey-esp committed:
Merge branch 'main' into b338873783-matrix-factorization

tswast reviewed (approved):
So cool! Thanks for your patience and persistence on this.
